designs/nodepool-requirement-flexibility.md

```diff
@@ -0,0 +1,185 @@
+
+This RFC proposes an additional key `minValues` in the NodePool requirements block, allowing Karpenter to be aware of user-specified flexibility minimums while scheduling pods to a cluster. If Karpenter cannot meet this minimum flexibility when scheduling a pod, it will fail the scheduling loop for that NodePool, either falling back to another NodePool which meets the pod requirements or failing scheduling the pod altogether.
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+        - key: karpenter.kwok.sh.instance-family
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2
+        - key: node.kubernetes.io/instance-type
+          operator: Exists
+          minValues: 10
+```
+
+## Use-Cases
+
+1. I want to configure Karpenter to enforce more flexibility when launching spot nodes, ensuring that I am constantly using enough instance types that I get a high-availability instance type
+2. I want to increase the number of nodes that Karpenter creates by ensuring that Karpenter doesn’t over-pack pods onto nodes when it’s scheduling
+3. I want to ensure that application pods scheduling to this NodePool *must* be flexible to different instance types, architectures, etc.
+
+## Solutions
+
+### [Recommended] Solution 1: `minValues` field in the `spec.requirements`  section of NodeClaim Template (becomes part of the template on the NodePool)
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+      - key: karpenter.kwok.sh.instance-family
+        operator: In
+        values: ["c", "m", "r"]
+        minValues: 2
+      - key: node.kubernetes.io/instance-type
+        operator: Exists
+        minValues: 10
+```
+
+We would add a new `minValues` field to the requirements section of the NodeClaim template. By adding this section to the NodeClaim template, we inform the scheduler that it should stop continuing to bin-pack pods onto Preflight NodeClaims and should, instead, create another Preflight NodeClaim that will schedule that pod.
+
+#### Pros/Cons
+
+1. 👍👍 Puts all core scheduling behavior together in the same block, allowing users to easily reason about what their requirements and minimums mean
+2. 👎 Piggy-backs off of a well-known core API. There’s risk that if this API changes in a way that is incompatible with this value in the future, we may have to break our `minValues` API
+
+### Solution 2: `minValues` in NodeClass API
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: EC2NodeClass
+spec:
+  minValues:
+    node.kubernetes.io/instance-type: 10
+    karpenter.kwok.sh.instance-family: 2
```

Comment:
how do defaults come into play for these values? requirements are defined by each individual cloud provider so an implementation outside of nodeclass becomes akward when talking about defaulting. The reason being you would need karpenter to be aware of the requirements keys. 

Alertnatively you could only allow specifying "core" requirements selectors which ALL Providers would need to implement, we all can share instance type, and instance family and cpu count and gpu count etc. 

Then you can default the casing for spot that for spot pools we want to default a certain number of instance types to avoid your thrashing case. 

I also like defaulting because we dont have to specify values everywhere, and the idea behind AKS's approach for nap is we are attempting to set smart defaults for the user. Ideally one `OOB` would get a great experience and wouldn't have to configure these values themselves. 

Karpenter should provide the flexibility to be more restrictive or have better control without burdening users with that same control.

Reply:
See now you talked about it in the cons